keps/sig-auth/3766-referential-authorization/README.md

```diff
@@ -82,59 +94,88 @@ been [used by
-  use cases. (It would be impossible to represent concepts like "all namespaces"
-  or label selectors that have become important for this KEP).
+* Advanced capabilities like label selectors and the ability to allow references
+  from any namespace are likely reasonable future extensions of this API but are
+  intentionally out of scope for the initial work.
 
 ## Proposal
 
-Move the existing ReferenceGrant resource into a new
-`authorization.k8s.io` API group, defined within the Kubernetes code base
-as part of the 1.27 release.
-
-We will take this opportunity to clarify and update the API after SIG Auth
-feedback. This resource will start with v1alpha1 as the API version.
-
+Create 3 new API types in a new `reference.authorization.k8s.io` API Group:
+
+1. **ClusterReferenceGrant:** This resource grants access to resources by
+   following references within the same namespace. For example, a `gateway-tls`
+   ClusterReferenceGrant would grant access to Gateway controllers for
+   same-namespace references from Gateways to TLS Secrets. API developers would
+   bundle these resources with their APIs.
+2. **ClusterReferenceConsumer:** This resource links a consumer (defined by
+   something like a RBAC Subject) to (Cluster)ReferenceGrants. For example, a
+   Gateway controller that used a ServiceAccount for auth, would
+   include a ClusterReferenceConsumer tying that ServiceAccount to references
+   from "Gateways" to "Secrets" for "tls". This would link that consumer to any
+   (Cluster)ReferenceGrants with the same `from`, `to`, and `for` values.
+3. **ReferenceGrant:** This resource authorizes specific instances of
+   cross-namespace references. For example, if an owner of a `acme-cert` Secret
+   in namespace `foo` wants to allow a TLS reference to that Secret from a
+   Gateway in namespace `bar`, they would create a ReferenceGrant that allowed
+   references from Gateways in the `bar` namespace to the `acme-cert` Secret in
+   their namespace (`foo`).
 
 ### Risks and Mitigations
 
-#### No Default Implementation
-Similar to the Ingress and Gateway APIs, this API will be dependent on
-implementations by controllers that are not included by default in Kubernetes.
-This could lead to confusion for users. We'll need to rely heavily on
-documentation for this feature, tracking all uses of official Kubernetes APIs
-that support ReferenceGrant in a central place.
-
 #### Potential for Variations Among Implementations
+Although the authorization logic for ReferenceGrant will be built into
+Kubernetes, controllers will still need to perform additional checks. For
+example, consider the case where 2 Gateways target the same Secret, but only
+one of those references is allowed by a ReferenceGrant. The authorization logic
+built into Kubernetes would have granted the controller access to the Secret
+but the controller would still need to perform separate checks to ensure that
+a ReferenceGrant had allowed a specific reference.
```

Comment:
Is there any way a controller could indicate in an request that it's following a reference (e.g. a "from" resource?), such that the authorizer could check that there is in fact a grant from that referrer resource to the target resource?

Reply:
That's definitely something that @deads2k and @liggitt have suggested, but it's unclear how that would work. I think it was @stealthybox that suggested something like impersonation here but that turned into a dead end IIRC. IMO this is something that could be added after the fact without actually changing the user-facing API. For now I think the current proposal solves ~90% of the problem and still leaves room for solving this part in the future (with the same API).